File: js/test-smoke.mjs
// Smoke tests for all 4 JS primitives. Run with: node js/test-smoke.mjs

import assert from "node:assert/strict";
import { parse, argsAfter, invoke, SkillNotFound } from "./slash-skill.js";
import { Plan, AlreadyInPlan, NotInPlan } from "./plan-mode.js";
import { shouldReview, gate, ReviewVerdict } from "./reviewer-gate.js";
import { ArtifactStore } from "./rlm.js";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

// 1. slash-skill
assert.equal(parse("/weather Tokyo"), "weather");
assert.equal(parse("/plan"), "plan");
assert.equal(parse("hello world"), null);
assert.equal(parse("/BadName"), null);
assert.equal(argsAfter("/weather Tokyo"), "Tokyo");
const v = await invoke("echo", { echo: () => "ok" });
assert.equal(v, "ok");
await assert.rejects(() => invoke("nope", {}), SkillNotFound);
console.log("  [1/4] slash-skill OK");

// 2. plan-mode
const p = new Plan("s1");
p.enter("migrate").write(["a", "b"]).exit(true);
assert.equal(p.approved, true);
assert.equal(p.mode, "build");
assert.deepEqual(p.todos, ["a", "b"]);
// Re-enter then try to enter again -> AlreadyInPlan
p.enter("second");
assert.throws(() => p.enter("third"), (e) => e instanceof AlreadyInPlan);
const p2 = new Plan("s2");
assert.throws(() => p2.write(["x"]), (e) => e instanceof NotInPlan);
p2.enter("g").write(["x"]).exit(false);
assert.equal(p2.approved, false);
console.log("  [2/4] plan-mode OK");

// 3. reviewer-gate
assert.equal(await shouldReview("hi"), false);
assert.equal(await shouldReview("x".repeat(500) + " see fig. 2"), true);
const cleanR = async (t) => new ReviewVerdict(true, [], "haiku");
const dirtyR = async (t) => new ReviewVerdict(false, ["fake ref"], "sonnet");
const v1 = await gate("x".repeat(500), [cleanR]);
assert.equal(v1.clean, true);
const v2 = await gate("x".repeat(500), [cleanR, dirtyR]);
assert.equal(v2.clean, false);
assert.deepEqual(v2.issues, ["fake ref"]);
await assert.rejects(() => gate("x", []), /at least one reviewer/);
console.log("  [3/4] reviewer-gate OK");

// 4. rlm
const tmp = mkdtempSync(join(tmpdir(), "openmneme-js-"));
try {
  const store = new ArtifactStore(tmp);
  const art = await store.register("s1", Buffer.from("col1,col2\n1,2\n"), "csv", "tiny table");
  assert.equal(art.bytes, 14);
  assert.ok(art.shortRef().includes("tiny table"));
  const data = await store.resolve("s1", art.id);
  assert.equal(Buffer.compare(data, Buffer.from("col1,col2\n1,2\n")), 0);
  const list = await store.list("s1");
  assert.equal(list.length, 1);
  await assert.rejects(() => store.resolve("s1", "nope"));
  const n = await store.prune(0);
  assert.ok(n >= 1);
} finally {
  rmSync(tmp, { recursive: true, force: true });
}
console.log("  [4/4] rlm OK");

console.log("\nALL 4 PRIMITIVES: SMOKE OK");

File: tests/test_smoke.py
"""Smoke tests for all 4 primitives. Run with: python3 tests/test_smoke.py"""

import asyncio
import shutil
import sys
import tempfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from openmneme import slash_skill, plan_mode, reviewer_gate, rlm


def test_slash_skill():
    assert slash_skill.parse("/weather Tokyo") == "weather"
    assert slash_skill.parse("/plan") == "plan"
    assert slash_skill.parse("hello world") is None
    assert slash_skill.parse("/BadName") is None
    assert slash_skill.args_after("/weather Tokyo") == "Tokyo"
    assert slash_skill.args_after("/plan") == ""

    async def run():
        reg = {"echo": lambda: "ok", "async_echo": lambda: asyncio.sleep(0, "async-ok")}
        assert await slash_skill.invoke("echo", reg) == "ok"
        assert await slash_skill.invoke("async_echo", reg) == "async-ok"
        try:
            await slash_skill.invoke("nope", reg)
        except slash_skill.SkillNotFound as e:
            assert e.name == "nope"
        else:
            raise AssertionError("expected SkillNotFound")

    asyncio.run(run())
    print("  [1/4] slash_skill OK")


def test_plan_mode():
    p = plan_mode.Plan(session_id="s1")
    p.enter("goal A").write(["a", "b"]).exit(True)
    assert p.approved and p.mode == "build" and p.todos == ["a", "b"]

    p.enter("goal B").write(["c"])
    try:
        p.enter("goal C")  # already in plan
    except plan_mode.AlreadyInPlan:
        pass
    else:
        raise AssertionError("expected AlreadyInPlan")

    p2 = plan_mode.Plan(session_id="s2")
    try:
        p2.write(["x"])  # not in plan
    except plan_mode.NotInPlan:
        pass
    else:
        raise AssertionError("expected NotInPlan")
    p2.enter("g").write(["x"]).exit(False)
    assert p2.approved is False and p2.mode == "build"
    print("  [2/4] plan_mode OK")


def test_reviewer_gate():
    assert asyncio.run(reviewer_gate.should_review("hi")) is False
    long_with_fig = "x" * 500 + " see fig. 2"
    assert asyncio.run(reviewer_gate.should_review(long_with_fig)) is True

    async def clean_rv(text):
        return reviewer_gate.ReviewVerdict(clean=True, issues=[], reviewer="haiku")

    async def dirty_rv(text):
        return reviewer_gate.ReviewVerdict(clean=False, issues=["fake ref"], reviewer="sonnet")

    v1 = asyncio.run(reviewer_gate.gate("x" * 500, [clean_rv]))
    assert v1.clean and v1.reviewer == "haiku"

    v2 = asyncio.run(reviewer_gate.gate("x" * 500, [clean_rv, dirty_rv]))
    assert not v2.clean and "fake ref" in v2.issues and v2.reviewer == "sonnet"

    try:
        asyncio.run(reviewer_gate.gate("x" * 500, []))
    except ValueError:
        pass
    else:
        raise AssertionError("expected ValueError for empty reviewers")

    print("  [3/4] reviewer_gate OK")


def test_rlm():
    with tempfile.TemporaryDirectory() as tmp:
        store = rlm.ArtifactStore(root=tmp)
        art = store.register("s1", b"col1,col2\n1,2\n", "csv", "tiny table")
        assert art.bytes == 14
        assert "tiny table" in art.short_ref()
        assert store.resolve("s1", art.id) == b"col1,col2\n1,2\n"
        assert store.resolve("s1", f"art:{art.id}") == b"col1,col2\n1,2\n"
        assert len(store.list("s1")) == 1

        try:
            store.resolve("s1", "nope")
        except rlm.ArtifactNotFound:
            pass
        else:
            raise AssertionError("expected ArtifactNotFound")

        # prune
        store.register("s1", b"old", "text")
        n = store.prune(ttl_days=0)
        assert n >= 1
    print("  [4/4] rlm OK")


if __name__ == "__main__":
    test_slash_skill()
    test_plan_mode()
    test_reviewer_gate()
    test_rlm()
    print("\nALL 4 PRIMITIVES: SMOKE OK")
